Attempt at providing an owned version of command

masklinn opened the pull request:
Pushing this to run the tests as getting things to work locally on a mac seems like a bit of a bear.

- Could not get the lifetimes right to not break `RemoteChild::session` (the lifetime becomes bound to the child's rather than the original session's), so move it into its own impl block, and added an extremely debatable `Child::into_session`, since it means not being able to call `wait`, or `wait_with_output`, or `disconnect`, so probably not useful (but a slightly separate variant of `Child::session` seems awful).
- Initially wanted to have a version directly owning a `Session`, but that turns out impossible as the `Command` builder works off of mutable references, so not possible to consume the builder in order to get the session out, short of creating either separate `impl` blocks for `Command<&Session>`, `Command<Arc<Session>>` and `Command<Session>`, or separate consuming spawns, neither sounding great.
- Naming is dubious, especially `shared_command`, but I don't recall ever seeing a method around shared refs so I went simple. I would have impl'd `Arc<Session>::into_command` but that's not legal...

  Maybe a free function would be better?

Hopefully will eventually fix #117

jonhoo commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/openssh-rust/openssh/136)
<!-- Reviewable:end -->

masklinn commented:
@NobodyXu can you approve workflow runs? Also maybe comment on this early starting point.

NobodyXu commented:
P.S. I don't think implementing `Command<Session>` is a good idea, since it's never designed to be taken by value.

`Command<Arc<Session>>` or `Command<Rc<Session>>` is more similar to `Command<&Session>` and I'd like to see support for them.

NobodyXu commented on src/child.rs:
```diff
@@ -74,18 +74,20 @@ macro_rules! delegate {
-impl<'s> RemoteChild<'s> {
+pub type RemoteChild<'a> = Child<&'a Session>; 
+
+impl<T> Child<T> {
```
```suggestion
impl<S: AsRef<Session> + Clone> Child<S> {
```

NobodyXu replied:
With the `Clone` bound, you can also implement `fn session() -> S`.

masklinn replied:
Yeah I figured that afterwards but didn't have the time to update the PR.

I don't think the `AsRef` bound is useful since at the end fo the day the `Client` doesn't actually care about the `Session`, beyond the outstanding reference in the borrow case. And the `Clone` case is only useful for `Session` itself.

I guess I could bound the entire thing anyway to ensure `Child` is only used with a session-adjacent object, but it feels unnecessary?

NobodyXu replied:
That's true, so it's probably ok for `Child` to not constraint `S` since it is not used.

NobodyXu commented on src/child.rs:
```diff
@@ -74,18 +74,20 @@ macro_rules! delegate {
 #[derive(Debug)]
-pub struct RemoteChild<'s> {
-    session: &'s Session,
+pub struct Child<T> {
+    session: T,
```
```suggestion
pub struct Child<S> {
    session: S,
```

NobodyXu commented on src/session.rs:
```diff
@@ -277,6 +277,15 @@ impl Session {
         )
     }
 
+    pub fn shared_command<'a, S: Into<Cow<'a, str>>>(self: std::sync::Arc<Self>, program: S) -> OwnedCommand<std::sync::Arc<Self>> {
```
I think `shared_thread_safe_command` is a better name.
Then we can have `shared_command` for `Rc<...>`

masklinn replied:
Added that change but TBH it feels unwieldingly verbose.

NobodyXu replied:
Hmmm, maybe arc_command is a better name?

NobodyXu commented on src/session.rs:
```diff
@@ -277,6 +277,15 @@ impl Session {
+        self.shared_raw_command(&*shell_escape::unix::escape(program.into()))
+    }
+
+    pub fn shared_raw_command<S: AsRef<OsStr>>(self: std::sync::Arc<Self>, program: S) -> OwnedCommand<std::sync::Arc<Self>> {
```
I would like to also have another generic version of this that takes `this: impl AsRef<Session> + Clone`

masklinn replied:
I'm not sure `impl` can be used this way, and I'm reasonably certain it it's not supported (yet) for a `self`. It could be a free or factory function, but in that case maybe it should be a public constructor for `OwnedCommand` instead?

masklinn replied:
[Currently valid list of self types](https://doc.rust-lang.org/reference/items/associated-items.html#methods) is `Self`, reference, box, rc, arc, and pin. So yeah no impl as an instance method.

NobodyXu replied:
You can make it an associated function instead `this: impl AsRef<Session> + Clone` is valid.

masklinn replied:
Oh I see that's what you meant, sorry I misunderstood.

Should it just replace the Arc version then? As it kinda obviates it, and if we call it "to_command" or something like that, even with an explicit `Arc` wrapper it's a lot shorter still.

NobodyXu replied:
You could keep it as a convenient method, or remove it if you think it's unnecessary.

masklinn replied:
I'll be honest here, I don't use the crate, I just saw the issue during jonhoo's stream and figured I had an idea which could solve the problem with not too much code duplication (or macro).

So while I do have my own sensibilities about code and APIs, I'd *much* rather go with what you, jon, and other actual users of the crate (maybe the other maintainers of openssh-sftp-client since that's where things started?) would prefer.

NobodyXu replied:
In that case, let's keep the arc_command, AFAIK many do use it from multi-threaded async context with Arc.

masklinn replied:
OK if it's a common use case it makes sense to have a convenient shortcut. And the associated function does look pretty good in the test.

I made the arc versions just be convenience shortcuts over the associated function as that seems simpler / clearer.

NobodyXu commented on tests/openssh.rs:
```diff
@@ -1057,3 +1057,40 @@ async fn test_read_large_file_bug() {
+
+#[tokio::test]
+#[cfg_attr(not(ci), ignore)]
+async fn rc_client() {
```
```suggestion
async fn arc_client() {
```

masklinn commented:
> P.S. I don't think implementing `Command<Session>` is a good idea, since it's never designed to be taken by value.

I was mostly thinking of the case where a user might want to run a single command on the session, but didn't want to bother with borrowing. However as I wrote up it doesn't work without changing the `Command` API entirely so that's moot.

NobodyXu commented on src/command.rs:
```diff
@@ -207,17 +207,19 @@ where
 #[derive(Debug)]
-pub struct Command<'s> {
-    session: &'s Session,
+pub struct OwnedCommand<S> {
```
Can we rename this to `CommandAnyClonable`?

masklinn replied:
Ah sorry I had not seen your comment before doing a bit more shuffling around, squashing the commits, and writing docs.

I ended up renaming the internal command back to `Command`, and doing the aliasing on the edge in `lib.rs`, with the inner generic `Command` being exported as `OwningCommand` (seems better than `Owned`), and the `openssh::RemoteChild` and `openssh::Command` aliases being created there (`lib.rs`) rather than inside the crate guts.

I can still rename `crate::command::Command` if you prefer some other name tho.

NobodyXu commented on src/session.rs:
```diff
@@ -270,13 +272,85 @@ impl Session {
     /// If `program` is not an absolute path, the `PATH` will be searched in an OS-defined way on
     /// the host.
-    pub fn raw_command<S: AsRef<OsStr>>(&self, program: S) -> Command<'_> {
+    pub fn raw_command<S: AsRef<OsStr>>(&self, program: S) -> Command<&'_ Self> {
```
You can forward command and raw_command to to_command and to_raw_command

masklinn replied:
Interestingly no, because there is no `AsRef<T> for &T`.

NobodyXu replied:
Hmmm, that's unexpected

masklinn replied:
Yep it also surprised me when I hit it. One of the reasons I'm always a little bit confused about using AsRef v Borrow v Deref.

Here AsRef might be the wrong choice because it allows "transverse" ref conversions e.g. `AsRef<OsStr> for str` and has more flexibility but I'm not sure that's needed.

`Borrow` and `Deref` are more limited but might be more suitable.

NobodyXu replied:
In that case, maybe swap that to Deref, since reference/Arc/Rc etc all have deref implemented.

NobodyXu commented on src/session.rs:
````diff
@@ -327,7 +401,7 @@ impl Session {
     /// # Ok(()) }
     /// ```
-    pub fn subsystem<S: AsRef<OsStr>>(&self, program: S) -> Command<'_> {
+    pub fn subsystem<S: AsRef<OsStr>>(&self, program: S) -> Command<&'_ Self> {
````
Can we also have a generic version for subsystem please?

It cannot be emulated by to_command or to_raw_command

masklinn replied:
Sure, should it have both the `to_` and the `arc` version, or just the `to_` version?

NobodyXu replied:
Just to to version will be good enough.

NobodyXu commented on src/command.rs:
```diff
@@ -167,29 +167,30 @@ impl<S> OverSsh for &mut S
 ///
-/// `Command` can be reused to spawn multiple remote processes. The builder methods change the
-/// command without needing to immediately spawn the process. Similarly, you can call builder
+/// `OwnedCommand` can be reused to spawn multiple remote processes. The builder methods change
```
Did you forget to update the name?

masklinn replied:
Yep, looks like I missed this one.

NobodyXu commented on src/lib.rs:
```diff
@@ -167,12 +167,16 @@ mod builder;
 
 mod command;
-pub use command::{Command, OverSsh};
+pub use command::{Command as OwningCommand, OverSsh};
```
I suggest renaming the type instead of creating a type alias, it would make our doc confusing to user

masklinn replied:
Damn and here I'd thought I could avoid renaming `Command` everywhere.

masklinn commented:
- moved the trait bound from AsRef to Deref
  - made `[raw_]command` go through `to_[raw_]command` since they now can
- renamed the owning `Command` to `OwningCommand` instead of just aliasing it as the crate root, for documentary clarity
- added a `to_subcommand` associated function to do the session-owning thing for subcommand features

NobodyXu commented on src/session.rs:
```diff
@@ -270,20 +270,89 @@ impl Session {
+    /// })
+    /// # Ok(()) }
+    pub fn arc_command<'a, P: Into<Cow<'a, str>>>(
+        self: std::sync::Arc<Session>,
```
I think taking a 

```suggestion
        self: &std::sync::Arc<Session>,
```

is better and more ergonomic, otherwise users would have to do `session.clone().arc_command(..)`

masklinn replied:
I think the current version is better, because otherwise the arc has to be cloned internally, so if the caller just wants to create a movable command/child they can't just move the one arc they have they get additional and unnecessary refcount traffic.

NobodyXu replied:
That's reasonable.

NobodyXu commented:
BTW, you need to fix the CI before I can merge this PR.

masklinn commented:
> BTW, you need to fix the CI before I can merge this PR.

Yep but since I can't run the tests locally I can only push and wait for CI approval. I'll get on that.

NobodyXu commented:
@masklinn There's a [`run_ci_tests.sh`](https://github.com/openssh-rust/openssh/blob/master/run_ci_tests.sh) in the repository that can be used to run some of the tests.

For linting, it's just `cargo-clippy` and `cargo-fmt`.

masklinn commented:
> There's a [run_ci_tests.sh](https://github.com/openssh-rust/openssh/blob/master/run_ci_tests.sh) in the repository that can be used to run some of the tests.

Yeah but it requires docker, and I'm on a mac and not super willing to install docker desktop.

> For linting, it's just `cargo-clippy` and `cargo-fmt`.

Yeah I ran fmt but it changed imports I had not touched so I figured the repo had strange configuration and reverted them, and they show up as lint errors so I guess it's just that they were last fmt'd before a change in configuration or something.

masklinn commented:
Hopefully CI is good now, I broke one of the existing doc comments by mistake, and the new `arc_command` example had a few brainfarts.

NobodyXu commented:
> Yeah but it requires docker, and I'm on a mac and not super willing to install docker desktop.

You can use [OrbStack](https://orbstack.dev/), an alternative to docker desktop that claims to be faster than docker desktop and uses less energy.

I use it for docker desktop now and it indeed startup quickly.

> It's nbd as far as I'm concerned just a bit more latency but I'm not replying within seconds anyway. Unless you have to approve every CI run I can imagine that gets old.

Yes, I do need to approve every CI run.

Once this PR lands in main, any new PR from you will have CI run automatically without having me to approve.

> so I guess it's just that they were last fmt'd before a change in configuration or something.

It's probably because new lints are added to newer cargo.

NobodyXu commented on tests/openssh.rs:
```diff
@@ -1057,3 +1057,73 @@ async fn test_read_large_file_bug() {
+
+#[tokio::test]
+#[cfg_attr(not(ci), ignore)]
+async fn generic_client() {
```
```suggestion
async fn rc_client() {
```

I think `rc_client` is a better name here.

masklinn replied:
Dunno, the intent of the name was to indicate that we're testing the generic `Session::to_command` rather than the ref-specific Session::command` or the Arc-specific `Session::arc_command`, the Rc is an implementation detail of the test but I figured I might as well use that since references and arc are already exercised. Would it help if I made most of the body generic and then ran it with both Rc and Arc?

NobodyXu replied:
It makes sense then, though I would like to rename this and the other test fn  to test_session_to_command and test_session_arc_command

masklinn replied:
Works for me.

NobodyXu commented:
The coverage failure is probably due to CI setup (e.g. GHA does not allow PRs from non-member to use secrets) or whatever, I can bypass that and merge the PR once you address my review.

NobodyXu commented:
Thank you @masklinn !

NobodyXu commented:
Released in v0.10.1
